web/pandas/pdeps/0008-inplace-methods-in-pandas.md

```diff
@@ -0,0 +1,403 @@
+
+For the second reason: we are convinced that this is not worth it. While it might save some keystrokes (if you have a
+long variable name), this code style also has sufficient disadvantages that we think it is not worth providing "two
+ways" to achieve the same result:
+
+- You can't use method chaining with `inplace=True`
+- The ``inplace`` keyword complicates type annotations (because the return value depends on the value of `inplace`)
+- Using `inplace=True` gives code that mutates the state of an object and thus has side-effects. That can introduce
+  subtle bugs and is harder to debug.
+
+Finally, there are also methods that have a `copy` keyword instead of an `inplace` keyword (which also avoids copying
+the data in the case of `copy=False`, but still returns a new object referencing the same data instead of updating the
+calling object), adding to the inconsistencies. This `copy=False` option also has become redundant with the introduction
+of Copy-on-Write.
+
+Given the above reasons, we are convinced that there is no need for neither the `inplace` nor the `copy` keyword (except
+for a small subset of methods that can actually update data inplace). Removing those keywords will give a more
+consistent and less confusing API.
+
+Thus, in this PDEP, we aim to standardize behavior across methods to make control of inplace-ness of operations
+consistent, and compatible with Copy-on-Write.
+
+Note: there are also operations (not methods) that work inplace in pandas, such as indexing (
+e.g. `df.loc[0, "col"] = val`) or inplace operators (e.g. `df += 1`). This is out of scope for this PDEP, as we focus on
+the inplace behaviour of DataFrame and Series _methods_.
+
+## Detailed description
+
+### Status Quo
+
+Many methods in pandas currently have the ability to perform an operation inplace. For example, some methods such
+as ``DataFrame.insert``, only support inplace operations, while other methods use keywords such as ``copy``
+or ``inplace`` to control whether an operation is done inplace or not.
+
+Unfortunately, many methods supporting the ``inplace`` keyword either cannot be done inplace, or make a copy as a
+consequence of the operations they perform, regardless of whether ``inplace`` is ``True`` or not. This, coupled with the
+fact that the ``inplace=True`` changes the return type of a method from a pandas object to ``None``, makes usage of
+the ``inplace`` keyword confusing and non-intuitive.
+
+In addition, some methods, such as ``DataFrame.rename`` and ``DataFrame.rename_axis`` confusingly support both
+the ``copy`` and ``inplace`` keywords, with the value of ``inplace`` overwriting the value of ``copy``.
+
+To summarize the status quo of inplace behavior of methods, we have divided methods that can operate inplace or have
+an ``inplace``/``copy`` keyword into 4 groups:
+
+**Group 1: Methods that always operate inplace**
+
+| Method Name   |
+|:--------------|
+| ``insert``    |
+| ``pop``       |
+| ``update``    |
```

Comment:
is `update` always inplace? e.g. in
```python
In [43]: df1 = pd.DataFrame({'numbers': [1, np.nan]})

In [44]: df2 = pd.DataFrame({'numbers': ['foo', 'bar']})

In [45]: df1.update(df2)
```

Reply:
It always operates inplace on the __pandas object__, but not necessarily on the underlying data. Your example would update df1, but the array would be copied before the update actually happens. Same as with replace and friends when the value to set is not compatible with the array dtype

Reply:
sure but then should it belong in group 2?

Reply:
There is no inplace keyword, so we don't have to change anything, same with insert and pop

Reply:
I don't understand the title then

> methods that always operate inplace

As in, methods that always operate inplace in the pandas object?

Because in another part of the tutorial the wording

> Some of the methods with an `inplace` keyword can actually work inplace

is used - I find it confusing when you're talking about "actually working inplace" and "modifying the pandas object"

Reply:
The problem is that "inplace" has multiple meanings (mutate object vs mutate values), and it's not easy to always correctly frame this. 
We can do a better job of first describing those two different meanings, but I don't know which terms we can use to make this clearer in every place where the word "inplace" is used.

Reply:
There are 2 different kinds of inplace (we might have to clarify this better)

- inplace on the pandas object, e.g. insert, group 1 is this
- inplace on the underlying array, e.g. modifying the values inplace. group 2 These method can (potentially) operate inplace, not guaranteed though

Update is a mixture of both, but for this proposal it belongs into group 1 since it always updates the pandas object inplace and _might_ update the underlying values inplace (but this is a side note here)

Reply:
We could do something like consistently use "object-inplace" and "values-inplace" instead of just "inplace" to clarify what is meant (if a specific type of inplaceness is meant). Would that help? (if we clearly define those two terms)

Reply:
Yes, that would make it clearer, thanks

Reply:
If we change the keywords do we just go all in and make those copy_pandas_object and copy_underlying_data? Or something similar. 

Obviously more verbose but yea I agree with @MarcoGorelli  sentiment that inplace is vague to the point of misinterpretation

Reply:
``copy_pandas_object`` won't be necessary with CoW. We will only need a keyword where we actually __can__ modify the underlying data inplace.

Reply:
@WillAyd to be clear, my comment above was about terminology to use in the text explaining this, not for actual keyword names (from your comment I get the impression you are talking about actual keywords)

Reply:
I'm probably way too abstracted from all of this, but find the terminology confusing. I didn't realize that CoW only ever referred to the pandas "container" object and didn't take a point of view on the underlying data. So if that' is the long term commitment then yea `copy_pandas_object` is a bit redundant, although I wonder if there is other terminology we need to cover if/when CoW extends to the underlying data

Generally it feels like there is a lot to be gained from a more explicit keyword than `inplace` when we want to modify the underlying data; I think is in line with @Dr-Irv points